results/cmd/api/main.go

```diff
@@ -89,53 +90,60 @@ func (s *server) CreateTaskRunResult(ctx context.Context, req *pb.CreateTaskRunR
-	}
-	taskrun := &pb.TaskRun{}
-	rowNum := 0
-	for rows.Next() {
-		var taskrunblob []byte
-		rowNum++
-		if rowNum >= 2 {
-			log.Println("Warning: multiple rows found")
-			break
-		}
-		rows.Scan(&taskrunblob)
-		if err := proto.Unmarshal(taskrunblob, taskrun); err != nil {
-			log.Fatal("unmarshaling error: ", err)
-			return nil, fmt.Errorf("failed to unmarshal taskrun: %w", err)
-		}
-	}
 	return &pb.TaskRunResult{TaskRun: taskrun, ResultsId: req.GetResultsId()}, nil
 }
 
 // UpdateTaskRun receives TaskRun and FieldMask from client and uses them to update records in local Sqlite Server.
 func (s *server) UpdateTaskRunResult(ctx context.Context, req *pb.UpdateTaskRunRequest) (*pb.TaskRunResult, error) {
-	// Update the entire row in database based on uid of taskrun.
-	statement, err := s.db.Prepare("UPDATE taskrun SET name = ?, namespace = ?, taskrunlog = ? WHERE results_id = ?")
+	// Find corresponding TaskRun in the database according to results_id.
+	tx, err := s.db.Begin()
 	if err != nil {
-		log.Printf("failed to update a existing taskrun: %v\n", err)
-		return nil, fmt.Errorf("failed to update a exsiting taskrun: %w", err)
+		log.Println("failed to begin a transaction: ", err)
+		return nil, fmt.Errorf("failed to update a taskrun: %w", err)
 	}
+	updateTaskRun, err := s.getTaskRunByID(req.GetResultsId())
+	if err != nil {
+		return nil, fmt.Errorf("failed to find a taskrun: %w", err)
+	}
+	// Merge TaskRun from client into existing TaskRun based on fieldmask.
 	taskrunFromClient := req.GetTaskRun()
-	blobData, err := proto.Marshal(taskrunFromClient)
+	fieldMask := req.GetUpdateMask()
+	// Update entire taskrun if user do not specify paths
+	if fieldMask == nil {
+		updateTaskRun = taskrunFromClient
+	} else {
+		msk, err := mask.FromFieldMask(fieldMask, updateTaskRun, false, true)
+		// Return NotFound error to client field is invalid
+		if err != nil {
+			log.Println("failed to convert fieldmask to mask: ", err)
+			return nil, status.Errorf(codes.NotFound, "field in fieldmask not found in taskrun")
+		}
+		if err := msk.Merge(taskrunFromClient, updateTaskRun); err != nil {
+			log.Println("failed to merge new taskrun into old taskrun: ", err)
+			return nil, fmt.Errorf("failed to update taskrun: %w", err)
+		}
+	}
+	blobData, err := proto.Marshal(updateTaskRun)
 	if err != nil {
 		log.Println("taskrun marshaling error: ", err)
 		return nil, fmt.Errorf("taskrun marshaling error: %w", err)
 	}
-	taskrunMeta := taskrunFromClient.GetMetadata()
+	statement, err := s.db.Prepare("UPDATE taskrun SET name = ?, namespace = ?, taskrunlog = ? WHERE results_id = ?")
+	if err != nil {
+		log.Printf("failed to update a existing taskrun: %v\n", err)
+		return nil, fmt.Errorf("failed to update a exsiting taskrun: %w", err)
+	}
+	taskrunMeta := updateTaskRun.GetMetadata()
 	if _, err := statement.Exec(taskrunMeta.GetName(), taskrunMeta.GetNamespace(), blobData, req.GetResultsId()); err != nil {
+		tx.Rollback()
```

Comment:
Do we need to rollback anywhere there is an error, or only on the exec?

Reply:
I think we only need to rollback only when we do update or any modification operations on database. Other errors do not damage the status of database